### ISSUE TITLE
ATP Cup

### ISSUE THREAD

**Issue by demod-au** [xbmc-catchuptv-au/plugin.video.catchuptv.au.nine#16]
I can't see any of the ATP-Cup "clips" in the kodi addon.  They do not show up as episodes in the json response - they show up as clips.  I've had a quick hack but can't seem to flush it out.  

They also show live matches, that are not always shown on the live channel.  Would be great to be able to see this tournament and also the Australian Open in a few weeks.  There's other tennis tournaments too that have the issue.  

Hopefully there is an easy way to add it.  I'm still trying to work out how the addon parses the data.

**Comment by demod-au** [comment]
Just created a pull request to add all live events to the live events page.  Still working on clips.  Bit more to get my head around :)

**Comment by glennguy** [comment]
Any 'clips' or 'extras' are currently filtered out. Someone asked me a while ago to add them in but I never got around to it.

I was thinking of having and option in the settings to toggle them on/off, and also maybe nesting them inside their own menu. I haven't looked at how the their data is organised but that will shape how it's best done.

Let me know if you need any help otherwise I could probably bang out something later today.

**Comment by demod-au** [comment]
I managed to get the "latest" clips to populate,  by using the episode url without "/episodes".  Which is fine for the tennis, but everything else just seems to duplicate the actual episodes if there are any.   Then I realised there is a "/clips" response that returns everything.  Might be overwhelming to just chuck it in with episodes. 

I ended up botching it with adding the latest clips (approx 7 of the newest), only if the genre is sport.  But some sports have episodes and they get duplicated too.  Maybe I need another check.  Not sure if this is worthy to go into the release.  

I'm not sure on what the best approach should be.  Maybe a clips menu item as you suggested.

**Comment by demod-au** [comment]
Ok, got a chance to have a look again.  Made a check for "Latest Clips".  function below. 
Seems to work well in that it only shows the latest clips if there are any.   

Not sure how to add the menu item.  The clip url is the same, just use /clips instead of /episodes.  It will be the same code below for clips to parse. 

```
def list_episodes(params):
    """
    Create and return list of episode objects
    """
    def get_metadata(episode):
        if not episode:
            return
        # filter extras again as some show are unable to be filtered at the
        # previous step
        if not episode.get('episodeNumber'):
            return
        # make sure season numbers match, some shows return all seasons.
        if episode['partOfSeason'].get('slug') != params['season_slug']:
            return

        e = classes.episode()
        e.episode_no = str(episode['episodeNumber'])
        e.thumb = episode['image']['sizes'].get('w480')
        e.fanart = data['tvSeries']['image']['sizes'].get('w1280')
        e.episode_name = episode.get('name').encode('utf8')
        e.title = e.get_title()
        e.desc = utils.ensure_ascii(episode.get('description'))
        e.duration = episode['video'].get('duration')//1000
        e.airdate = episode.get('airDate')
        e.id = episode['video'].get('referenceId')
        e.drm = episode['video'].get('drm')
        return e

    def get_clipmetadata(episode):
        e = classes.episode()
        e.episode_no = 'CLIP'
#str(episode['episodeNumber'])
        e.thumb = episode['image']['sizes'].get('w480')
        e.fanart = data['tvSeries']['image']['sizes'].get('w1280')
        e.episode_name = episode.get('name').encode('utf8')
#        e.title = 'CLIP:'+ e.get_title()
        e.title = e.get_title()
        e.desc = utils.ensure_ascii(episode.get('description'))
        e.duration = episode['video'].get('duration')//1000
        e.airdate = episode.get('airDate')
        e.id = episode['video'].get('referenceId')
        e.drm = episode['video'].get('drm')
        return e

    listing = []

    url = config.EPISODEQUERY_URL.format(
        params['series_slug'], params['season_slug']+'/episodes')
    data = fetch_url(url)
    for episode in data['episodes'].get('items'):
        e = get_metadata(episode)
        if e:
            listing.append(e)

    url = config.EPISODEQUERY_URL.format(
        params['series_slug'], params['season_slug'])
    data = fetch_url(url)
#    if data['genre'].get('name') == 'Sport':
    if data['items'][0].get('title') == 'Latest Clips': 
#    add menu item for "All Clips"
        for episode in data['items'][0].get('items'):
            e = get_clipmetadata(episode)
            if e:
                listing.append(e)

    return listing
```

**Comment by demod-au** [comment]
Anyone have an idea on how to create a menu item at this location?   I have all the scrapers ready to go, just not sure how to link the menu, with my functions.

Not sure if it is of much use,  I used the clips for a little bit while the ATP Cup was on, but there hasn't been much else that actually uses it.   The "live" changes in the pull request were gold during the Australian Open.  

Happy to close if not much interest.  I'll keep the "latest clips" in my local copy.  Not sure its worth adding the above to the main code?

**Comment by glennguy** [comment]
I think the least cluttering and easiest way would be to append another list item to the Sports genre 'series list' in menu.make_series_list after it's done processing the main list so it's just another folder along the other series. Pop it in at line 117 after the for block, just make a new ListItem and append it. And maybe remove the addSortMethod call so the list stays in the same order with the clips folder at the bottom?

You'll need to give that list item a similar url, set the title param to 'Latest Clips' or something similar
i.e `url = '{0}?action=listseries&title=Latest&20Clips'.format(_url)`
Only thing to watch out here is I can't remember if the %20 will be unescaped in default.py when the folder is entered... 

In comm.py alter what you have above to set `url` based on params['title'] - if 'Latest Clips' then use the /clips path so it's either doing episodes or clips, not both. This is where you'll need to check if %20 has been converted back to a space, if it's easier just remove the space altogether.

I might have forgotten something but I'm hopeful that will get it going for you!

This entire addon needs a bit of love and it's on my radar - Kodi 19 is probably not long from beta and we need to get it python3 compatible. iView and all of the sports addons are done, just this one and sbs needs to be sorted.

**Comment by demod-au** [comment]
Alrighty..  thanks for those tips!  put me on the right track.  I accidentally "upgraded" my local kodi copy and lost all my existing work! Had to start from scratch again. 

This was how I ended up doing it:

- you get to see clips in a series ONLY if there are "latest clips" (turns out nearly all shows/events don't have any!)
- It will load the 10 Latest clips - after episodes. (Turns out there are either episodes or clips, can't say I found one with both)
- it will add a menu item for "All Clips" - which loads all available clips for the series - 49 for the ATP. 

Just in time hey :) 

I'll attach the 3 files I modified here.  Probably doing it wrong?  Can do a pull request if it looks ok.  

[comm.py.txt](https://github.com/aussieaddons/plugin.video.catchuptv.au.nine/files/4338140/comm.py.txt)
[menu.py.txt](https://github.com/aussieaddons/plugin.video.catchuptv.au.nine/files/4338141/menu.py.txt)
[default.py.txt](https://github.com/aussieaddons/plugin.video.catchuptv.au.nine/files/4338143/default.py.txt)
